src/expr.rs

```diff
@@ -297,120 +289,142 @@ fn unary_op(input: (&[u8], EvalResult)) -> Option<EvalResult> {
     }
 }
 
-macro_rules! numeric (
-	($i:expr, $submac:ident!( $($args:tt)* )) => (map_opt!($i,$submac!($($args)*),EvalResult::as_numeric));
-	($i:expr, $f:expr ) => (map_opt!($i,call!($f),EvalResult::as_numeric));
-);
+fn numeric<I: Clone, E: nom::error::ParseError<I>, F>(
+    f: F,
+) -> impl Fn(I) -> nom::IResult<I, EvalResult, E>
+where
+    F: Fn(I) -> nom::IResult<I, EvalResult, E>,
+{
+    nom::combinator::map_opt(f, EvalResult::as_numeric)
+}
 
 impl<'a> PRef<'a> {
-    method!(unary<PRef<'a>,&[Token],EvalResult,crate::Error>, mut self,
-		alt!(
-			delimited!(p!("("),call_m!(self.numeric_expr),p!(")")) |
-			numeric!(call_m!(self.literal)) |
-			numeric!(call_m!(self.identifier)) |
-			map_opt!(pair!(one_of_punctuation!(["+", "-", "~"]),call_m!(self.unary)),unary_op)
-		)
-	);
+    fn unary<'t>(&'_ self, input: &'t [Token]) -> CResult<'t, EvalResult> {
```

Comment:
Please write `&self` instead of `&'_ self` (everywhere)

Reply:
I _think_ it's actually important for it to be `'_` so that the compiler does not assume that the lifetime of `&self` is also included in `'t`. We want the returned value to be able to outlive `&self`. But it could be that my understanding of inferred lifetime bounds is incorrect here?

Reply:
There are special rules for `&self` if you don't manually specify any lifetime parameters. But an unspecified lifetime parameter is never assumed to be the same as a specified one.